### ISSUE TITLE
Appearance cost has no effect

### ISSUE THREAD

**Issue by mpandoko** [mikel-brostrom/Yolov5_DeepSort_OSNet#374]
in [deepsort tracker](https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/master/deep_sort/sort/tracker.py)

`# Now Compute the Appearance-based Cost Matrix
        app_cost = self.metric.distance(
            np.array([dets[i].feature for i in detection_indices]),
            np.array([tracks[i].track_id for i in track_indices]),
        )`

Why is line 121 fetch the track_id instead of the feature of the track?

It seems to be wrong since the appearence cost always return way higher than the threshold
Edit: The problem is not in the track_id, but in the distance function. See below comments

**Comment by HeChengHui** [comment]
I went to check the [original repos](https://github.com/nwojke/deep_sort/blob/280b8bdb255f223813ff4a8679f3e1321b08cdfc/deep_sort/tracker.py) implementation and looks like it was also done this way.
Line 98 is the cost matrix calculation. in `distance(features, targets)`, the code for `targets` also seems to use track_id (Line 97). Can double confirm in case I am reading it wrong.

**Comment by mpandoko** [comment]
I think I found the real problem. In [nn_matching](https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/master/deep_sort/sort/nn_matching.py), line 91, this repos uses compute distance matrix from torchreid. However, the original repos uses def _cosine_distance that normalize the appearence features first.

Secondly, is there any reason this repos removed [i, :] in line 165?

**Comment by HeChengHui** [comment]
> I think I found the real problem. In [nn_matching](https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/master/deep_sort/sort/nn_matching.py), line 91, this repos uses compute distance matrix from torchreid. However, the original repos uses def _cosine_distance that normalize the appearence features first.

https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/680062c8221c6a247f319dee7ec36537329ca6e1/deep_sort/sort/nn_matching.py#L91

In distance.py:
`def compute_distance_matrix(input1, input2, metric='euclidean'):`

since metric is cosine:
`distmat = cosine_distance(input1, input2)`

```
def cosine_distance(input1, input2):
    """Computes cosine distance.
    Args:
        input1 (torch.Tensor): 2-D feature matrix.
        input2 (torch.Tensor): 2-D feature matrix.
    Returns:
        torch.Tensor: distance matrix.
    """
    input1_normed = F.normalize(input1, p=2, dim=1)
    input2_normed = F.normalize(input2, p=2, dim=1)
    distmat = 1 - torch.mm(input1_normed, input2_normed.t())
    return distmat
```

looks like it is normalised to me.

**Comment by mpandoko** [comment]
You are correct. Before continuing, may I ask, do you able to reproduce the problem? (That is, the app_cost is not in the range of 0-1, but in the scale of tens or hundreds)

In this repos, I could not find that the app_cost specified the metrics being used. Is it possible the app_cost actually uses euclidean distance, instead of cosine distance?

**Comment by HeChengHui** [comment]
> In this repos, I could not find that the app_cost specified the metrics being used. Is it possible the app_cost actually uses euclidean distance, instead of cosine distance?

If you are using the default deepsort model i believe it uses euclidean distance as stated in the [model zoo](https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO) under same domain ReID.

**Comment by jckhng** [comment]
> I think I found the real problem. In [nn_matching](https://github.com/mikel-brostrom/Yolov5_DeepSort_Pytorch/blob/master/deep_sort/sort/nn_matching.py), line 91, this repos uses compute distance matrix from torchreid. However, the original repos uses def _cosine_distance that normalize the appearence features first.
> 
> Secondly, is there any reason this repos removed [i, :] in line 165?

I think [i,:] is required. The cost_matrix is a MxN matrix and not a row vector. If you run the code, it will just return cost_matrix as a single row. The code will still run, but the appearance cost will not have any effect. 

> > In this repos, I could not find that the app_cost specified the metrics being used. Is it possible the app_cost actually uses euclidean distance, instead of cosine distance?
> 
> If you are using the default deepsort model i believe it uses euclidean distance as stated in the [model zoo](https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO) under same domain ReID.

The default model is osnet_ibn_x1_0_MSMT17 which uses euclidean for metric (the code also assume euclidean for default), according to the model zoo. I am not able to verify if it is indeed the case. However, the default threshold is [0.2](https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/e263d94ddf9b4d5f30426a1eac508b335f5abb77/deep_sort/deep_sort.py#L23), which fails almost everything [here](https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/e263d94ddf9b4d5f30426a1eac508b335f5abb77/deep_sort/sort/tracker.py#L123) measured in euclidean. 

Suggested changes to fix the re-id performance:
1. Change
https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/e263d94ddf9b4d5f30426a1eac508b335f5abb77/deep_sort/deep_sort.py#L51-L52
to 
```
        metric = NearestNeighborDistanceMetric(
            "cosine", max_cosine_distance, nn_budget)
```

2. Change 
https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/e263d94ddf9b4d5f30426a1eac508b335f5abb77/deep_sort/sort/nn_matching.py#L162-L164
to 
```
        for i, target in enumerate(targets):
            cost_matrix[i, :]= self._metric(self.samples[target], features)
        return cost_matrix
```

**Comment by HeChengHui** [comment]
@jckhng Thanks for the investigating further. will try out when there is time.

> I think [i,:] is required. The cost_matrix is a MxN matrix and not a row vector. If you run the code, it will just return cost_matrix as a single row. The code will still run, but the appearance cost will not have any effect.

could this be why changing the reid model have no changes to the output? I only see the time taken changes.

**Comment by mikel-brostrom** [comment]
Hi @jckhng , @mpandoko ! Sorry for my late reply.

> The default model is osnet_ibn_x1_0_MSMT17 which uses euclidean for metric (the code also assume euclidean for default), according to the model zoo. I am not able to verify if it is indeed the case.

Yes, euclidean is the metric for this model as can be seen here: https://kaiyangzhou.github.io/deep-person-reid/MODEL_ZOO

This:

```python
for i, target in enumerate(targets):
            cost_matrix[i, :]= self._metric(self.samples[target], features)
        return cost_matrix
```

is correct. I must have introduced the bug in one of my latest updates. Thanks for pointing this out!

Here:

```python
metric = NearestNeighborDistanceMetric(
            "cosine", max_cosine_distance, nn_budget)
```

it makes more sense to have a max_distance for cosine and one for euclidean.

**Comment by mpandoko** [comment]
Thank you for your reply, also thank you for making this awesome repos.

> it makes more sense to have a max_distance for cosine and one for euclidean.

The problem is that the calculation of appearance cost here

https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/da3b55546671dab6e5e6592cf89dd7e71ab0eaee/deep_sort/sort/tracker.py#L118-L123

will use the euclidean distance instead of the cosine distance it's supposed to. And the app_cost will always return a value much bigger than the appearance matching threshold, making no matches at all in this stage and only relying on the IoU matching.

**Comment by mikel-brostrom** [comment]
Yes, that is why it makes more sense to have suitable matching threshold for `cosine` and one for the `euclidean` distance

**Comment by mpandoko** [comment]
As to my understanding from the paper, app_cost should use the `cosine`. But here, it uses `euclidean` in the appearance distance calculation.

And also, would you like to verify this statement
> app_cost will always return a value much bigger than the appearance matching threshold

Thank you

**Comment by mikel-brostrom** [comment]
> As to my understanding from the paper, app_cost should use the cosine. But here, it uses euclidean in the appearance distance calculation.

Yes, I am using euclidean distance as it is the default for the appearance models as described here https://kaiyangzhou.github.io/deep-person-reid/pkg/engine.html?highlight=euclidean

> app_cost will always return a value much bigger than the appearance matching threshold

Yes, this is because the euclidean distance was not normalized. I am working on this as well.

**Comment by mikel-brostrom** [comment]
Master just got an update. It should be fixed now. If you run this under `deep_sort/sort/tracker.py`

```python
# Now Compute the Appearance-based Cost Matrix
        app_cost = self.metric.distance(
            np.array([dets[i].feature for i in detection_indices]),
            np.array([tracks[i].track_id for i in track_indices]),
        )
        print('app_cost\n', app_cost)
        app_gate = app_cost > self.metric.matching_threshold
        print('pos_cost\n', pos_cost)
        # Now combine and threshold
        cost_matrix = self._lambda * pos_cost + (1 - self._lambda) * app_cost
        print('cost_matrix\n', cost_matrix)
        cost_matrix[np.logical_or(pos_gate, app_gate)] = linear_assignment.INFTY_COST
        print('cost_matrix\n', cost_matrix)
```

You should see something like this (with `lambda=0`):

```bash
0: 480x640 1 person, 1 cell phone, Done. YOLO:(0.036s), DeepSort:(0.016s)
app_cost
 [[   0.039068     0.94478]
 [    0.92121    0.035968]]
pos_cost
 [[   0.011589      1.6839]
 [     2.5068    0.013374]]
cost_matrix
 [[   0.039068     0.94478]
 [    0.92121    0.035968]]
cost_matrix
 [[   0.039068       1e+05]
 [      1e+05    0.035968]]
0: 480x640 1 person, 1 cell phone, Done. YOLO:(0.038s), DeepSort:(0.016s)
app_cost
 [[   0.036643     0.91375]
 [     1.0216    0.045099]]
pos_cost
 [[   0.026824      1.6863]
 [     2.4653    0.017884]]
cost_matrix
 [[   0.036643     0.91375]
 [     1.0216    0.045099]]
cost_matrix
 [[   0.036643       1e+05]
 [      1e+05    0.045099]]
0: 480x640 1 person, 1 cell phone, Done. YOLO:(0.037s), DeepSort:(0.017s)
app_cost
 [[   0.018242     0.92245]
 [    0.99623    0.046107]]
pos_cost
 [[  0.0085211      1.6636]
 [     2.4598    0.057298]]
cost_matrix
 [[   0.018242     0.92245]
 [    0.99623    0.046107]]
cost_matrix
 [[   0.018242       1e+05]
 [      1e+05    0.046107]]
```

previously the output looked like this:

```bash
0: 480x640 1 person, 1 cell phone, Done. YOLO:(0.038s), DeepSort:(0.018s)
app_cost
 [     1163.8      12.233]
pos_cost
 [[   0.027182      2.0376]
 [     3.2275    0.020425]]
cost_matrix
 [[     1163.8      12.233]
 [     1163.8      12.233]]
cost_matrix
 [[      1e+05       1e+05]
 [      1e+05       1e+05]]
0: 480x640 1 person, 1 cell phone, Done. YOLO:(0.038s), DeepSort:(0.018s)
app_cost
 [     1187.2      22.609]
pos_cost
 [[  0.0032156      2.0147]
 [     3.2493    0.023461]]
cost_matrix
 [[     1187.2      22.609]
 [     1187.2      22.609]]
cost_matrix
 [[      1e+05       1e+05]
 [      1e+05       1e+05]]
0: 480x640 1 person, 1 cell phone, Done. YOLO:(0.038s), DeepSort:(0.018s)
app_cost
 [     1170.2      8.7373]
pos_cost
 [[  0.0072399      2.0278]
 [     3.2263    0.026806]]
cost_matrix
 [[     1170.2      8.7373]
 [     1170.2      8.7373]]
cost_matrix
 [[      1e+05       1e+05]
 [      1e+05       1e+05]]
```

You can see that everything makes much more sense now.

**Comment by mpandoko** [comment]
> Yes, I am using euclidean distance as it is the default for the appearance models as described here https://kaiyangzhou.github.io/deep-person-reid/pkg/engine.html?highlight=euclidean
> 
> Yes, this is because the euclidean distance was not normalized. I am working on this as well.

Now I see. Thank you!

**Comment by jckhng** [comment]
@mikel-brostrom 

https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/f2ae9213b79360cb27e641eb85c605b523ff8b9e/deep_sort/sort/nn_matching.py#L69-L70

We should revisit this code. I think that normalizing x and y vector and then calculating the distance between these normalized vectors will no longer give the euclidean distance between x and y vector that the function is supposed to return. Reverting to the previous iteration of code for these two lines should make more sense.

**Comment by mikel-brostrom** [comment]
Euclidean is no longer used as distance metric. Check:

https://github.com/mikel-brostrom/Yolov5_DeepSort_OSNet/blob/f2ae9213b79360cb27e641eb85c605b523ff8b9e/deep_sort/deep_sort.py#L52

Reason is:

https://github.com/KaiyangZhou/deep-person-reid/issues/502

**Comment by mikel-brostrom** [comment]
Thank you @KaiyangZhou  for taking time for answering our questions!
